test/e2e/framework/job/wait.go

```diff
@@ -27,8 +28,15 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/utils/format"
+	"k8s.io/utils/ptr"
 )
 
+// JobState is used to verify if Job matches a particular condition.
+// First parameter means that the condition was found.
+// Return false and an error where notify caller that condition was not found.
+type JobState func(job *batchv1.Job) (bool, error)
```

Comment:
I still think we should simplify this. From the comment it is not clear whether `true, non-nil-err` is valid (it is, but is treated the same as `false, non-nil-err` and thus makes no sense). `false, nil` is also valid, but not very useful: `WaitForJobState` then falls back to "job did not match intended condition function", which conveys no information about what the function was checking for.

How about this?

```
// JobState is used to verify if Job matches a particular condition.
// If it matches, an empty string is returned.
// Otherwise, the string explains why the condition is not matched.
// This should be a short string. A dump of the job object will
// get added by the caller.
type JobState func(job *batchv1.Job) string
```

Then below `WaitForJobState` uses:
```
framework.MakeMatcher(func(job *batchv1.Job) (func() string, error) {
			why := state(job)
			if why == "" {
			        nil, nil
		        }
			return func() string {
				     return fmt.Sprintf("%v\n%s", why, format.Object(job, 1))
	                }, nil
```

BTW, as far as I'm concerned you can always squash immediately, as long as you don't rebase.